src/Type/Php/CallUserFuncArrayDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,38 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\NeverType;
+use PHPStan\Type\Type;
+
+class CallUserFuncArrayDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'call_user_func_array';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		$args = $functionCall->getArgs();
+		$defaultReturn = ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+
+		if (count($args) <= 1) {
+			return new NeverType();
+		}
+
+		$callbackType = $scope->getType($args[0]->value);
+		if ($callbackType->isCallable()->yes()) {
+			return $scope->getType(new FuncCall($args[0]->value, array_slice($args, 1)));
```

Comment:
It's suspicious to me that both of these extensions are implemented identically although `call_user_func_array` and `call_user_func` behave differently.

i think that `call_user_func_array` has wrong implementation because `call_user_func_array('foo', [1, 2, 3])` will lead to `('foo')([1, 2, 3])` but it should lead to `('foo')(1, 2, 3)`.

In case of call_user_func_array you have to account for:

1) General arrays
2) Constant arrays
3) Array unpacking, including `call_user_func_array(...$args)` and `call_user_func_array('foo', ...$args)`.